.github/workflows/deploy.yaml

```diff
@@ -0,0 +1,65 @@
+name: Deploy
+
+on:
+  push:
+    branches: [ main ]
+
+  workflow_dispatch:
+
+concurrency: production
+
+jobs:
+  build:
+    uses: ./.github/workflows/test.yaml
+    with:
+      upload-artifact: true
+
+  deploy:
+    runs-on: ubuntu-latest
+    environment: production
+
+    steps:
+      - name: Pre-deploy checks
+        uses: actions/github-script@v6
+        with:
+          script: |
+            // Check that the code to be deployed is the latest code on the branch on GitHub
+            // If you want to rollback changes to prod, consider using a revert commit to main
+            //
+            const response = await github.rest.repos.getCommit({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              ref: 'refs/heads/main',
+            });
+            const mainBranchCommitSha = response.data.sha
+            if (context.sha !== mainBranchCommitSha) {
+              core.info(`Workflow run GITHUB_SHA ${context.sha} does not match main branch commit sha ${mainBranchCommitSha}`)
+              core.error('Code to deploy does not match code in main branch')
+              core.setFailed('Refusing to deploy, please do not re-run this workflow run')
+            }
+
+      - name: Download build artifact
+        uses: actions/download-artifact@v3
```

Comment:
I'd be tempted to pin this to the SHA rather than the version `actions/download-artifact@fb598a63ae348fa914e94cd0ff38f362e927b741` in this case as the version tag can be replaced, but the SHA can't